game-core/src/main/java/games/strategy/triplea/ui/MovePanel.java

```diff
@@ -1147,41 +1149,33 @@ private void updateUnitsThatCanMoveOnRoute(final Collection<Unit> units, final R
       best = CollectionUtils.getMatches(best, Matches.unitIsNotSea());
     }
     sortUnitsToMove(best, route);
-    Collections.reverse(best);
-    List<Unit> bestWithDependents = addMustMoveWith(best);
-    final MoveValidationResult allResults;
-    getData().acquireReadLock();
-    try {
-      final MoveDescription move =
-          new MoveDescription(bestWithDependents, route, unitsToTransports, dependentUnits);
-      allResults =
-          AbstractMoveDelegate.validateMove(
-              moveType, move, getCurrentPlayer(), nonCombat, getUndoableMoves(), getData());
-    } finally {
-      getData().releaseReadLock();
-    }
+    best = Lists.reverse(best);
+    Tuple<MoveValidationResult, List<Unit>> result = validateMove(best, unitsToTransports, route);
+    final MoveValidationResult allResults = result.getFirst();
+    List<Unit> bestWithDependents = result.getSecond();
     MoveValidationResult lastResults = allResults;
     if (!allResults.isMoveValid()) {
       // if the player is invading only consider units that can invade
       if (!nonCombat
           && route.isUnload()
           && Matches.isTerritoryEnemy(getCurrentPlayer(), getData()).test(route.getEnd())) {
         best = CollectionUtils.getMatches(best, Matches.unitCanInvade());
-        bestWithDependents = addMustMoveWith(best);
-        final MoveDescription move =
-            new MoveDescription(bestWithDependents, route, unitsToTransports, dependentUnits);
-        lastResults =
-            AbstractMoveDelegate.validateMove(
-                moveType, move, getCurrentPlayer(), nonCombat, getUndoableMoves(), getData());
+        result = validateMove(best, unitsToTransports, route);
+        bestWithDependents = result.getSecond();
+        lastResults = result.getFirst();
       }
       while (!best.isEmpty() && !lastResults.isMoveValid()) {
-        best = best.subList(1, best.size());
-        bestWithDependents = addMustMoveWith(best);
-        final MoveDescription move =
-            new MoveDescription(bestWithDependents, route, unitsToTransports, dependentUnits);
-        lastResults =
-            AbstractMoveDelegate.validateMove(
-                moveType, move, getCurrentPlayer(), nonCombat, getUndoableMoves(), getData());
+        // Try to find a valid sublist, skipping equivalent units.
+        final Unit firstSkipUnit = best.get(0);
+        int startIndex = 1;
+        while (startIndex < best.size()
+            && unitsAreEquivalent(firstSkipUnit, best.get(startIndex))) {
+          startIndex++;
+        }
```

Comment:
I have a hard time understanding why your new code is supposed to be equivalent.

Why can "equivalent" units be skipped, and what makes them "equivalent" in this scenario?